openadapt/evaluators/evaluator.py

```diff
@@ -0,0 +1,204 @@
+
+    def __init__(
+        self,
+        model_name: str = "gpt2",
+        max_input_size: int = DEFAULT_MAX_INPUT_SIZE,
+        max_tokens: int = DEFAULT_MAX_TOKENS,
+        max_screen_size: Tuple[int, int] = DEFAULT_MAX_SCREEN_SIZE,
+    ):
+        self.model_name = model_name
+        self.max_input_size = max_input_size
+        self.max_tokens = max_tokens
+        self.max_screen_size = max_screen_size
+
+    @abstractmethod
+    def init_model(self):
+        """Initialize the model. This must be adapted for each model, given attribute model name"""
+        self.model = None
+
+    @abstractmethod
+    def get_completion(
+        self,
+        prompt: str,
+    ):
+        """Get the completion for a given prompt"""
+        return ""
+
+    def build_prompt(
+        self, ref_window: Window, action: KeyAction | MouseAction, active_window: Window
+    ):
+        """Build the prompt for the model. This must be adapted for each model.
+
+        Args:
+            ref_window (Window): reference window
+            action (KeyAction | MouseAction): action to be performed
+            active_window (Window): active window
+
+        Returns:
+            str: prompt for the model
+        """
+        prompt = (
+            f"{ref_window.dict()=} {action.dict()=} {active_window=} # Provide valid"
+            " Python3 code containing the action dicts by completing the following,"
+            " and nothing else: active_action_dict="
+        )
+        return prompt
+
+    def evaluate_single_action(self) -> float:
+        """Evaluate the model completion on a set of fixtures
+
+        Returns:
+            float: average score
+        """
+        self.init_model()
+        # [TODO] run evaluation on different fixtures
+        ref_window, action, active_window = fixtures.generate_single_mouse()
+        prompt = self.build_prompt(ref_window, action, active_window)
+        completion = self.get_completion(prompt)
+        score = self._score_completion(completion, action)
+        logger.info(f"Average score is {score=}")
+        return score
+
+    def _euclidean_distance(
+        self, point1: Tuple[float, float], point2: Tuple[float, float]
+    ):
+        """Calculate the euclidean distance between two points
+
+        Args:
+            point1 (Tuple[float, float]): first point
+            point2 (Tuple[float, float]): second point
+
+        Raises:
+            ValueError: if either point is not 2D
+
+        Returns:
+            float: euclidean distance between the two points
+        """
+        if len(point1) != 2 or len(point2) != 2:
+            raise ValueError("Both points must be 2D coordinates (x, y)")
+
+        x1, y1 = point1
+        x2, y2 = point2
+        distance = math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)
+        return distance
+
+    def _score_valid_actions(
+        self,
+        ref_action: KeyAction | MouseAction,
+        prediction_action: KeyAction | MouseAction,
+    ) -> float:
```

Comment:
Params we need : 
`ref_action`:
`ref_window`:
`active_window`:
`predicted_action:`

Criteria for `MouseAction`:

From `ref_action` and `ref_window`, we need to figure out the control location (bounding box) 

Then, for the given `active_window`, figure out if `predicted_action` correctly click or moved mouse to the control location (bounding box). If the ROI does not exist in the `active_window`, return `None`.

In the end, the output of the evaluation will be : `True`, `False` or `None`.


-